File: perf-results/cross-lang/benchmarks/binarytrees.lua
local function make_tree(depth)
  if depth > 0 then
    depth = depth - 1
    return { make_tree(depth), make_tree(depth) }
  end
  return { false, false }
end

local function check_tree(node)
  if node[1] then
    return 1 + check_tree(node[1]) + check_tree(node[2])
  end
  return 1
end

local N = tonumber(arg and arg[1]) or 10
local min_depth = 4
local max_depth = math.max(min_depth + 2, N)
local stretch_depth = max_depth + 1

io.write(string.format("stretch tree of depth %d\t check: %d\n", stretch_depth, check_tree(make_tree(stretch_depth))))

local long_lived_tree = make_tree(max_depth)

for depth = min_depth, max_depth, 2 do
  local iterations = 2 ^ (max_depth - depth + min_depth)
  local check = 0
  for i = 1, iterations do
    check = check + check_tree(make_tree(depth))
    check = check + check_tree(make_tree(depth))
  end
  io.write(string.format("%d\t trees of depth %d\t check: %d\n", iterations * 2, depth, check))
end

io.write(string.format("long lived tree of depth %d\t check: %d\n", max_depth, check_tree(long_lived_tree)))

File: perf-results/cross-lang/benchmarks/binarytrees.java
// binarytrees — idiomatic single-threaded Java, matched to the Python/Ruby reference
// Usage: java binarytrees <depth>

public class binarytrees {
    static final class TreeNode {
        TreeNode left, right;
        TreeNode(TreeNode left, TreeNode right) { this.left = left; this.right = right; }
    }

    static TreeNode makeTree(int depth) {
        if (depth == 0) return new TreeNode(null, null);
        depth -= 1;
        return new TreeNode(makeTree(depth), makeTree(depth));
    }

    static int checkTree(TreeNode node) {
        if (node.left == null) return 1;
        return 1 + checkTree(node.left) + checkTree(node.right);
    }

    public static void main(String[] args) {
        int n = args.length > 0 ? Integer.parseInt(args[0]) : 10;
        int minDepth = 4;
        int maxDepth = Math.max(minDepth + 2, n);
        int stretchDepth = maxDepth + 1;

        System.out.println("stretch tree of depth " + stretchDepth +
            "\t check: " + checkTree(makeTree(stretchDepth)));

        TreeNode longLivedTree = makeTree(maxDepth);

        for (int depth = minDepth; depth <= maxDepth; depth += 2) {
            int iterations = 1 << (maxDepth - depth + minDepth);
            int check = 0;
            for (int i = 1; i <= iterations; i++) {
                check += checkTree(makeTree(depth));
                check += checkTree(makeTree(depth));
            }
            System.out.println((iterations * 2) + "\t trees of depth " + depth +
                "\t check: " + check);
        }

        System.out.println("long lived tree of depth " + maxDepth +
            "\t check: " + checkTree(longLivedTree));
    }
}
